<hr style="border: 2px solid green;" />

Hello, Esther!🤚 

My name is Kseniya Vishnevskaya. I'm going to review your project 😊 

✏️You can find my comments in <font color='green'>green</font>, <font color='orange'>yellow</font> or <font color='red'>red</font> boxes like this:

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    I'm using green color if everything is done perfectly.</div>       
<br>
<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    if I can give some recommendations</div>   
<br>
<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    if the block requires some corrections. Work can't be accepted with the red comments</div>
 
Further in the file you can see my comments, try to take them into account when performing the next tasks. **Please, don't delete them.** :)

Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> In the blue boxes you can leave comments and ask questions.</div>


<hr style="border: 2px solid green;" />

### <font color='orange'>General feedback</font>

<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    <s>Thank you for submitting your project 🙂  I'm glad to say that you executed your project really well. I especially liked the style, code formatting, detailed conclusions and function writing. But there are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.</div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    There are some comments for further improvement. They don't mean that you did anything wrong, but they're things we wanted to point out that can help you further develop your skills. 

And also I leave several links that can help you to make your further projects even better:

* Text formatting:
<a>https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd</a>

*  How to deal with missing data:
<a>https://towardsdatascience.com/how-to-deal-with-missing-data-in-python-1f74a9112d93</a>

*  How to edit text in jupyter notebook using Markdown:
<a>https://www.earthdatascience.org/courses/intro-to-earth-data-science/file-formats/use-text-files/format-text-with-markdown-jupyter-notebook/</a> 
    
   
</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> Hi Kseniya! Thank you so much for reviewing the project! Thanks for the feedback and comments. I will try to respond and correct any comment.</div>

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
    Thanks for sending in your project with corrections 😊. It's clear you've put a lot of effort into it. You've done a really good job! Glad to say that your project  has been accepted.  Good luck on the next sprint! :) </div>   

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

Description of the data:
* <strong><em> children </em></strong> : the number of children in the family
* <strong><em> days_employed </em></strong> : how long the customer has been working
* <strong><em> dob_years </em></strong> : the customer’s age
* <strong><em> education </em></strong> : the customer’s education level
* <strong><em> education_id </em></strong> : identifier for the customer’s education
* <strong><em> family_status </em></strong> : the customer’s marital status
* <strong><em> family_status_id </em></strong> : identifier for the customer’s marital status
* <strong><em> gender </em></strong> : the customer’s gender
* <strong><em> income_type </em></strong> : the customer’s income type
* <strong><em> debt </em></strong> : whether the customer has ever defaulted on a loan
* <strong><em> total_income </em></strong> : monthly income
* <strong><em> purpose </em></strong> : reason for taking out a loan

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
   <s> Well done for not forgetting the title and description of the project task :). But it is also important to include a description of the data to be worked with.</div> 

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> I agree with you. But even that I did not write :(. It was already embedded in the project notebook.
I have added the values I will work with...</div>

<div style="border:solid green 2px; padding: 20px"> <b>Success 2:</b><br>
    👍</div>  

## Open the data file and have a look at the general information. 

In [2]:
import pandas as pd
credit_scor = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_scor.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> I tried to refer from here to a comment you wrote to me in section <strong><em> 1.1. Conclusions </em></strong> </div>

# Checking the data by columns:

In [3]:
#Check how many unique values there are in children column
credit_scor['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [4]:
#Since there are illogical values, check what percentage they constitute of all people
illogical_values = ((76+47)/21525)*100
illogical_values

0.5714285714285714

Test result:
* In the 'children' column there are some with 20 children and some with -1 child. These values together constitute 0.57% of the total values.

In [5]:
#Check where missing values are
credit_scor.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Test result:
* Columns 'days_employed' and 'total_income' contain NaN values

In [6]:
#Check the type of each column
credit_scor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Test result:
* In the 'days_employed' column the value appears in float even though days should be an integer.
* Some cells contain negative values (from looking at the table).
* Sometimes the number of working days is greater than a person's age (from looking at the table).

In [7]:
#Check for illogical values (eg age under 18 with children)
credit_scor.groupby('dob_years')['children'].sum()

dob_years
0      68
19      0
20     12
21     39
22     45
23     98
24    119
25    164
26    264
27    350
28    333
29    430
30    494
31    538
32    484
33    562
34    592
35    581
36    543
37    605
38    523
39    464
40    511
41    415
42    453
43    321
44    309
45    281
46    249
47    172
48    158
49    181
50    188
51    123
52    101
53     83
54     90
55     62
56    144
57     58
58     42
59     84
60     49
61     50
62     51
63     18
64     41
65      6
66      8
67      5
68      6
69     23
70      5
71      1
72      3
73      1
74      0
75      0
Name: children, dtype: int64

Test result:
* Some of those who are 0 years old, have children

In [9]:
#Check the different levels of education
credit_scor['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

Test result:
* In the 'education' column has a total of 5 types of education, but since there are entries with uppercase and lowercase letters it is considered that there are many more types of education.

In [10]:
#Check the different types of purpose
credit_scor['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
housing                                     647
purchase of the house                       647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
building a property                         620
purchase of my own house                    620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

Test result:
* The 'purpose' column has many types of purposes, but if we collect all the purposes we can see that in total there are only 4 purposes written in different variations (wedding, car, education, housing).

In [11]:
#Check for repeated rows
credit_scor.duplicated().sum()
credit_scor[credit_scor.duplicated()]

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
4182          1            NaN         34    BACHELOR'S DEGREE             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
7808          0            NaN         57  secondary education             1   
8583          0            NaN         58    bachelor's degree             0   
9238          2            NaN         34  secondary education             1   
9528          0            NaN         66  secondary education             1   
9627          0            NaN         56  secondary education             1   
10462         0            NaN         62  secondary education             1   
10697         0            NaN         40  secondary education             1   
10864         0            NaN         62  secondary education             1   
10994         0            NaN         62  secondary education             1   
11791         0            NaN         47  secondary education             1   
12373         0            NaN         58  secondary education             1   
12375         1            NaN         37  secondary education             1   
12736         0            NaN         59  secondary education             1   
13025         1            NaN         44  secondary education             1   
13639         0            NaN         64  secondary education             1   
13773         0            NaN         35  secondary education             1   
13878         1            NaN         31  secondary education             1   
13942         0            NaN         44  secondary education             1   
14432         2            NaN         36    bachelor's degree             0   
14832         0            NaN         50  secondary education             1   
15091         0            NaN         58  secondary education             1   
15188         0            NaN         60  secondary education             1   
15273         0            NaN         57  secondary education             1   
16176         0            NaN         47  secondary education             1   
16378         0            NaN         46  secondary education             1   
16902         2            NaN         39  secondary education             1   
16904         1            NaN         32    bachelor's degree             0   
17379         0            NaN         54    bachelor's degree             0   
17755         1            NaN         43  secondary education             1   
17774         1            NaN         40  secondary education             1   
18328         0            NaN         29    bachelor's degree             0   
18349         1            NaN         30    bachelor's degree             0   
18428         0            NaN         64  secondary education             1   
18521         0            NaN         56  secondary education             1   
18563         0            NaN         54  secondary education             1   
18755         0            NaN         58  secondary education             1   
19041         0            NaN         56  secondary education             1   
19184         0            NaN         46  secondary education             1   
19321         0            NaN         23  secondary education             1   
19387         0            NaN         38    bachelor's degree             0   
19688         0            NaN         61  secondary education             1   
19832         0            NaN         48  secondary education             1   
19946         0            NaN         57  secondary education             1   
20116         0            NaN         57  secondary education             1   
20165         0            NaN         42  secondary education             1   
20702  

Test result:
* There are 54 repeating rows

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>

*  Best practice: for convenience, it is common to keep all libraries used in code in a separate cell.
Please visit this link, there is a collection of interesting notebooks:
<a>https://gist.github.com/ocoyawale/54d92fd4bf92508a2a6e482b5fa480fd</a>
        
* I see that the numbering of the code cells does not start from 1. Before submitting the project, it is best to restart the notebook to check that all code cells are running correctly (`Kernel > Restart & Run All on the toolbar`).
    
</div>   

### Summary of problem

1. In the 'children' column there are some with 20 children and some with -1 child, using the method: credit_scor['children'].value_counts()

2. Columns 'days_emplpyed' contain NaN values, using the method: credit_scor.isnull().sum()

3. Three more problems in column 'days_employed': *Some cells contain negative values. *Sometimes the number of working days is greater than a person's age. *In the 'days_employed' column the value appears in float even though days should be an integer, using the method: credit_scor.info()

4. In the 'dob_year' column there are people aged 0, although some have children, it does not make sense, using the method credit_scor.groupby('dob_years')['children'].sum()

5. In the 'education' column has a total of 5 types of education, using the method: credit_scor['education'].value_counts() , but since there are entries with uppercase and lowercase letters it is considered that there are many more types of education.

6. Column 'total_income' contain NaN values, using the method: credit_scor.isnull().sum()

7. The 'purpose' column has many types of purposes using the method: credit_scor['purpose'].value_counts() , but if we collect all the purposes we can see that in total there are only 4 purposes written in different variations.

8. There are duplicates rows, using the method: credit_scor.duplicated().sum()

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    You've done a great job of sorting out the data description and writing detailed conclusions. Well done! 😊</div>    

<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
  <s>  But it is better to run the code from your output in the module where you analyse the data, rather than writing it as text in the output, because different datasets can run in the same notebook and it is important to have a full set of checks. Please, fix it. </div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> I'm not sure I understood that comment. I will try to write here what I understood: Do you actually want to see all the tests I did to find out the problems?
If that is indeed the answer then I tried to make it above. </div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks 2:</b><br>
    Yes, that's right, you did well to show the checks you did to identify problems in the data. But what I meant was that it is not formally accepted to write code in the conclusions, because the conclusions are most often read by customers who don't understand anything about code.  </div>  

## Data preprocessing

### Processing missing values

1. Since there are only 123 people with a number of children does not make sense, using the method: credit_scor['children'].value_counts() , it's about half a percent, I decided to delete these people from the list. 
*The reason why there are ilogic number of children can be due to misleading data entry, where 20 can be 2 and -1 can be one child.

In [12]:
credit_scor.drop(credit_scor[credit_scor['children'] == 20].index, inplace=True)
credit_scor.drop(credit_scor[credit_scor['children'] == -1].index, inplace=True)

<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
   <s> I agree that these values are wrong and should be deleted but in that case it is worth checking what percentage of this data is in the dataframe. </div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> Agree with you. I tried to check the percentage it above. Do you have a more elegant method? </div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks v2:</b><br>
    I meant something like this: <code>len(credit_scor[credit_scor['children'] == 20])/ len(credit_scor['children']) </code>
    This is to avoid accidentally deleting a large amount of data (usually not more than 5%)</div>   

2. & 3. Despite the many problems there are in the days_employed column, it does not need its data for further analysis. I will not delete it because they may need it for other purposes.
I will change negative values to positive ones.
I will arrange it by filling in the missing values on mean or median days.
*The reason there are negative values can be due to misleading data entry. The reason there are missing values could be either because the person has never worked or because he does not want to disclose this information to the bank.

In [13]:
#Change negative values to positive ones
credit_scor['days_employed'] = credit_scor['days_employed'].abs()

In [14]:
#Filling the missing values with mean or median days
mean_days_employed = credit_scor['days_employed'].mean()
median_days_employed = credit_scor['days_employed'].median()

#Both results do not make much sense
#The average shows that a person works for about 260 years, the median shows a job of 10 years

In [15]:
#I decided to fill the missing values on average days depending on the age of the person.
#At first I find the average number of working days depending on age
age_dependent_work = credit_scor.groupby('dob_years')['days_employed'].median()

#Then I fill the missing values with 0, so after that I will fill in the average of working days
credit_scor['days_employed'] = credit_scor['days_employed'].fillna(0)

#A function that returns the average working days depends on the age of the employee
def total_days_employed(row):
    dob_years = row['dob_years']
    days_employed = row['days_employed']
    
    if days_employed == 0:
        return age_dependent_work.loc[row['dob_years']]
    return days_employed
        
credit_scor['total_days_employed'] = credit_scor.apply(total_days_employed, axis=1)

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    It's great that you wrote a function to fill in the missing values of average days employed depending on the age of the person 👍 </div> 

6. It is difficult to know how to fill in the missing values of the 'total_income' column. Since filling them in on average is not fair because they may not really have income, on the other hand, it could be that this value was not entered by mistake and therefore, putting them to zero is also incorrect because they may have income in another bank. Therefore, they should be questioned and as soon as they have to check if they deserve a loan they will have to find out directly in front of them if they have income or not.

In [16]:
#Change the missing values of total_income to -1
credit_scor['total_income'] = credit_scor['total_income'].fillna(-1)

### Conclusion

I tried to fill in missing entries without almost hurting the table data and how it would affect future analysis.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    You have done well in processing missing values. I like that you make assumptions and justify them logically. </div> 

### Data type replacement

2. & 3. Replace data type of total_days_employed from real numbers to integers. The type of total_days_employed is float and I replace to int.

In [17]:
#Change the type of days to an integer
credit_scor['total_days_employed'] = credit_scor['total_days_employed'].astype(int)

4. Since according to the rest of the data it seems that the people whose age is 0, this is a mistake (for example because some of them have children or are married). I will change the age to the average age.

In [18]:
#At first find the average age
dob_years_ave = credit_scor['dob_years'].mean()

#Then insert the average for all values that are 0 in dob_years column
credit_scor.loc[credit_scor['dob_years'] == 0, 'dob_years'] = dob_years_ave
credit_scor['dob_years'] = credit_scor['dob_years'].astype(int)

5. Change all uppercase to lowercase letters in the 'education' column.

In [19]:
credit_scor['education'] = credit_scor['education'].str.lower()

### Conclusion

Modify and replace data according to problems so that the table can then be analyzed in an orderly manner.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Well, it's all good here. 🙂 </div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    Please visit this link, there is information on how to optimise data types and downsize them: 
    <a>https://vincentteyssier.medium.com/optimizing-the-size-of-a-pandas-dataframe-for-low-memory-environment-5f07db3d72e</a>
</div> 

### Processing duplicates

8. I have found repetitive rows using the method: credit_scor.duplicated().sum(), therefore, I remove the repeating rows so as not to affect the analysis of the table.
*The reason why there are repetitive rows can be due to the fact that they entered the rows twice by mistake or that the customer contacted the bank on two different occasions.

In [20]:
credit_scor.drop_duplicates(inplace=True)
credit_scor.duplicated().sum()

0

<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    <s>Please, don't forget to run the code to count the number of duplicates :) </div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> Absolutely right. I added the run code above </div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
    It's great that you checked for duplicates after deletion. I see that in line 9 you counted the number of duplicates, but the calculation overlapped with the table that you then output. It is important to see this number 54, because the number of duplicates can change if you use a different dataset. </div> 

### Conclusion

If there was an identification detail such as an ID number or name and date of birth, it could be said with certainty that these are repetitive rows and that it is the same person. At the same time, the chance that there are two people with the exact same data in the table with 2100 people is small, so the chance is that these are repetitive rows.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   Good guess :) </div>

### Categorizing Data

7. Because there are supposedly many reasons for taking a loan, it can be seen that they all have several keywords that are related to the same topic and therefore each category can be put into a certain category. I will use the Lemmatization method to put each purpose into its own category. I preferred the Lemmatization method because it is more accurate and advanced.

In [21]:
import nltk
from nltk.stem import WordNetLemmatizer
import pandas as pd
wordnet_lemma = WordNetLemmatizer()
def purpose_category(purpose):
    words=nltk.word_tokenize(purpose)
    lemmas=[wordnet_lemma.lemmatize(w, pos='n') for w in words]
    for lemma in lemmas:
        if lemma == 'wedding':
            return 'wedding_expenses'
        if lemma in ['car', 'cars']:
            return 'car_expenses'
        if lemma in ['estate', 'housing', 'house', 'property']:
            return 'housing_expenses'
        if lemma in ['university', 'education', 'educated']:
            return 'education_expenses'
credit_scor['purpose_category'] = credit_scor['purpose'].apply(purpose_category)

In order to determine who has children and who does not have children, I will put the children's column in a function that returns yes or no, depending on the number of children.

In [22]:
#Yes or no children category
def children_yes_no(children):
    if children >= 1:
        return 'yes'
    return 'no'
credit_scor['children_yes_no'] = credit_scor['children'].apply(children_yes_no)

In order to determine what the income level is, one must understand how many people there are in each household and then divide the total income by the number of people and understand what the income level is.

In [23]:
#number of people in a household
def number_of_people(row):
    children = row['children']
    family_status_id = row['family_status_id']
    if family_status_id == 0:
        return 2 + children
    if family_status_id == 1:
        return 2 + children
    if family_status_id == 2:
        return 1 + children
    if family_status_id == 3:
        return 1 + children
    if family_status_id == 4:
        return 1 + children
    
credit_scor['number_of_people'] = credit_scor.apply(number_of_people, axis=1)

In order to determine the level of income I use the data of the total income and divide by the total number of people. The level of income is determined approximately by the level of per capita income in Israel.

In [24]:
#income level category
def income_level(row):
    total_income = row['total_income']
    number_of_people = row['number_of_people']
    
    if 0 <= total_income / number_of_people <= 3000:
        return 'low income'
    if 3001 <= total_income / number_of_people <= 8000:
        return 'lower middle income'
    if 8001 <= total_income / number_of_people <= 15000:
        return 'upper middle income'
    if total_income / number_of_people >= 15001:
        return 'high income'
    return 'unknonw'

credit_scor['income_level'] = credit_scor.apply(income_level, axis=1)

### Conclusion

I divided into categories according to logic. At the same time, a division into other categories (such as income_level) can lead to different results, so in order to know for sure how to make the division I must rely on the bank's criteria for what is considered high or low income and whether there is reference to the number of persons in the household.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   I really liked your idea of segmenting bank customers by income. You took into account the composition of the family, which helped you know the average income per person.  </div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    To categorise the income, I can also suggest using the <code>describe()</code> method, it will indicate the quartile boundaries, which helps to divide the data into equal parts. And equal parts are more convenient to compare with each other :) </div> 

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> Grate! I did'nt know this method, it could have helped me a lot. I wanted to use it now but I realized it would change some of my conclusions so I will save it for next time. Thanks! </div>

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
   😊 </div>

In [25]:
credit_scor.groupby('income_level').agg({'income_level':'count'})

income_level
income_level                     
high income                  6280
low income                    369
lower middle income          5438
unknonw                      2094
upper middle income          7150

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [26]:
pivot_children = credit_scor.pivot_table(index = 'children_yes_no', values = 'debt', aggfunc = ['count', 'sum', 'mean'])
pivot_children

count   sum      mean
                  debt  debt      debt
children_yes_no                       
no               14091  1063  0.075438
yes               7240   669  0.092403

In [52]:
pivot_children1 = pd.pivot_table(credit_scor, index = 'children', columns = 'debt', values='family_' aggfunc = ['count', 'sum', 'mean'], margins=True)
pivot_children1

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000
All       21331  1732  0.081196

It can be seen that there is a relative relationship between having children or not and repaying a loan on time. People who have children are almost 2% more behind in repaying the loan than those who do not have children.

### Conclusion

There is a connection between children and the ability to repay a loan on time. But since there is no significant difference it is necessary to take into account additional parameters.

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b>

* To avoid making additional calculations for the debtors' share column, I can give you a hint on how to generate a summary table quickly. The <code>columns</code> parameter should be omitted, the <code>values</code> parameter should be set to the debt column, and the <code>aggfunc</code> parameter should be set to the function list ['count', 'sum', 'mean']. The result will be three columns: the total number of borrowers, the number of debtors and the proportion of debtors.
    
* You could also make a summary table, grouped by the number of children, to see how the number of children affects the loan repayment on time.
 </div> 

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b><br>

* Wow, I love it! I did not know that it can be used in such a way, it looks much better!

* Agree with you, but some people had a really small amount of a certain number of children (e.g. only a few individuals had five children) so it was not unequivocal enough to determine if the amount of children affects the ability to repay a loan so I preferred not to get into it.
</div>

- Is there a relation between marital status and repaying a loan on time?

In [25]:
pivot_family_status = credit_scor.pivot_table(index = 'family_status', values = 'debt', aggfunc = ['count', 'sum', 'mean'])
pivot_family_status

count  sum      mean
                    debt debt      debt
family_status                          
civil partnership   4134  385  0.093130
divorced            1189   84  0.070648
married            12261  927  0.075606
unmarried           2796  273  0.097639
widow / widower      951   63  0.066246

It can be seen that there is a connection between family status and the ability to repay a loan, with widows / widowers having the least difficulty repaying a loan on time, while unmarried and civil partnership have the most difficulty repaying a loan on time.

### Conclusion

Here, too, there is a connection between family status and the ability to repay a loan on time, but additional parameters need to be taken into account.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   Great!  </div>

- Is there a relation between income level and repaying a loan on time?

In [26]:
pivot_income_level = credit_scor.pivot_table(index = 'income_level', values = 'debt', aggfunc = ['count', 'sum', 'mean'])
pivot_income_level

count  sum      mean
                     debt debt      debt
income_level                            
high income          6280  477  0.075955
low income            369   30  0.081301
lower middle income  5438  482  0.088636
unknonw              2094  170  0.081184
upper middle income  7150  573  0.080140

It is almost impossible to deduce which income levels indicate the ability to repay on time, the percentages are really similar at all income levels.

### Conclusion

It may be that there is really no connection between the level of income and the ability to repay, but before such a conclusion can be drawn, one must check whether these are really the criteria for deciding which income indicates the level of income.

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    Please note that the low-income group is very small in volume, i.e. there is not enough data to reliably estimate the percentage obtained. That is why for such a group it is better to say that there is not enough data to estimate it. That is why I suggested above to divide into equal numbers of groups to make it easier to compare the percentages. </div> 

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b><br>
Thanks for the important comment! I will consider it from now on.
</div>

- How do different loan purposes affect on-time repayment of the loan?

In [27]:
pivot_purpose_category = credit_scor.pivot_table(index = 'purpose_category', values = 'debt', aggfunc = 
                                                 ['count', 'sum', 'mean'])
pivot_purpose_category

count  sum      mean
                     debt debt      debt
purpose_category                        
car_expenses         4279  400  0.093480
education_expenses   3988  369  0.092528
housing_expenses    10751  780  0.072551
wedding_expenses     2313  183  0.079118

It can be seen that those whose purpose is to purchase a car or education, it is more difficult for them to repay the loan on time.

### Conclusion

Here too the conclusion is not unequivocal sience the differences are not significant and additional parameters need to be considered. But if we just look at it, the purpose of the loan may be serious, buying a car is a luxury and people who take out a loan to buy a car are a little irresponsible. People who take out a loan for education are usually students who have difficulty repaying a loan because they work less.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   Yes, I totally agree with you :)  </div>

## General conclusion

* **Original conclusion:**

As I have already mentioned, the conclusions from the data analysis are not unequivocal since each time I referred to one parameter in relation to another parameter. In order to draw more accurate conclusions it is necessary to combine several parameters and then the analysis will be more accurate. For example, to score on a parameter and then in the end each customer gets a certain score considering all the parameters and it will be easier to decide whether to give him a loan and under what conditions.

* **Conclusion after changes:**

In this project, a dataset was obtained containing personal information regarding the bank's customers. The aim of the project was to preprocessing the data and arrange it in such a way that it could be analyzed afterwards.
Using different methods, missing values were found and problems were discovered in the dataset.

The problems discovered were illogical number of children, illogical number of working days in relation to the employee's age, some of the working days were given in negative numbers and most of them were of the float type. In addition, it was necessary to correct the different levels of education and change the values to lower case, categorize the popurpose column into several main purpose that contain all the types of purpose mentioned, and overcome the problem of repeatative rows.

In some cases it was necessary to decide what could be disposed of and which data were more or less significant for the analysis, and in some cases it was difficult to decide what could be done with the data without checking it with the provider who gave the data. 

After arranging the data and within the existing limitations, we moved on to the phase of analyzing the data in order to allow the bank to determine certain criteria of customers who are at risk of not repaying the loan on time. In this project, one criterion is examined each time, which affects the client's ability to repay a loan on time. Among the relationships examined were the relationship between the level of income and the ability to repay on time, people with or without children, marital status and the purpose of taking the loan.

In general it can be seen (above) that these parameters are related to the customer's repayment ability, for example different purposes for taking a loan causes different risks in its repayment ability, or, people with children will have more difficulty in repaying the loan compared to people without children. However, at the same time, when the customer comes to the bank to receive a loan, it is necessary to weigh all the data in order to determine the situation of each customer individually.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   That's right, a more comprehensive assessment needs a broader toolkit. But since this is our first project, we are working with the knowledge we have in reserve. There are more exciting adventures ahead 😉  </div>

<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
  <s> The final conclusion should be a brief overview of your work, so it's better to describe everything you learned about the data. Please write more about what you researched and what results you got in the process, as the final conclusion is the first thing people read and is often the basis for judging the quality of the work as a whole. </div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b><br>
Totally agree. I tried to write a broader conclusion based on your comment ...
</div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
    Well done, you you wrote detailed conclusions. Thanks a lot for your work! 🙂 But in analytics, it is also very important to back up your words with numbers.  </div>   


<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b><br>
Thanks so much for all the comments! I took to my attention and hope I understood them all.
As for the recommendations, I copied the links you referred me to and will slowly learn more about the "wonders" of notebook:)
</div>

## Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.